librz/analysis/p/analysis_bf.c

```diff
@@ -7,35 +8,446 @@
 #include <rz_asm.h>
 #include <rz_analysis.h>
 
-static size_t countChar(const ut8 *buf, int len, char ch) {
-	int i;
-	for (i = 0; i < len; i++) {
-		if (buf[i] != ch) {
-			break;
-		}
-	}
-	return i;
-}
-
 static int getid(char ch) {
 	const char *keys = "[]<>+-,.";
 	const char *cidx = strchr(keys, ch);
 	return cidx ? cidx - keys + 1 : 0;
 }
 
+/* New IL uplift bf */
+#define BF_ADDR_SIZE  64
+#define BF_ALIGN_SIZE 8
+
+struct bf_stack_t {
+	ut64 stack[32];
+	int sp;
+};
+typedef struct bf_stack_t *BfStack;
+
+typedef struct bf_context_t {
+	BfStack stack;
+	ut64 op_count;
+} BfContext;
+
+static void bf_syscall_read(RzILVM vm, RzILOp op) {
+	ut8 c = getc(stdin);
+	BitVector bv = rz_il_bv_new_from_ut32(BF_ALIGN_SIZE, c);
+
+	RzILVal ptr_val = rz_il_dump_value(rz_il_hash_find_val_by_name(vm, "ptr"));
+
+	rz_il_vm_mem_store(vm, 0, ptr_val->data.bv, bv);
+	rz_il_free_value(ptr_val);
+}
+
+static void bf_syscall_write(RzILVM vm, RzILOp op) {
+	RzILVal ptr_val = rz_il_dump_value(rz_il_hash_find_val_by_name(vm, "ptr"));
+
+	BitVector bv = rz_il_vm_mem_load(vm, 0, ptr_val->data.bv);
+	ut32 c = rz_il_bv_to_ut32(bv);
+
+	rz_il_free_value(ptr_val);
+	rz_il_bv_free(bv);
+
+	putchar(c);
+}
+
+ut64 pop_astack(BfStack stack) {
+	if (stack->sp < 0) {
+		printf("Empty Stack\n");
+		return -1;
+	}
+
+	ut64 ret = stack->stack[stack->sp];
+	stack->sp -= 1;
+	return ret;
+}
+
+void push_astack(BfStack stack, ut64 id) {
+	stack->sp += 1;
+	stack->stack[stack->sp] = id;
+}
+
+ut64 alloc_id(BfStack stack, ut64 addr) {
+	push_astack(stack, addr);
+	return addr;
+}
+
+RzPVector *bf_right_arrow(RzILVM vm, ut64 id) {
+	// (set ptr (+ (val ptr) (int 1)))
+	RzILOp var = rz_il_new_op(RZIL_OP_VAR);
+	var->op.var->v = "ptr";
+	var->op.var->ret = 0;
+
+	RzILOp int_ = rz_il_new_op(RZIL_OP_INT);
+	int_->op.int_->length = BF_ADDR_SIZE;
+	int_->op.int_->value = 1;
+	int_->op.int_->ret = 1;
+
+	RzILOp add = rz_il_new_op(RZIL_OP_ADD);
+	add->op.add->x = 0;
```

Comment:
Did we choose to keep the *index* behaviour? I still think that's a bit too risky, but I'm also ok to refactor this as next step. I understand there's a lot of code here.

Reply:
I hope "next step" doesn't mean after merging to dev.

Reply:
What is the index behavior? What do you mean? Not sure I understand cc @Heersin

Reply:
https://im.rizin.re/rizinorg/pl/9yycu548kfdbuqqwwczzsokpmc